bpf/controller.go

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2017 Kinvolk GmbH
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bpf
+
+import (
+	"context"
+	"time"
+
+	bpf "github.com/iovisor/gobpf/elf"
+)
+
+/*
+#include <linux/bpf.h>
+*/
+import "C"
+
+const (
+	packetsKey uint32 = iota
+	bytesKey
+)
+
+type Controller struct {
+	cgroup string
+	module *bpf.Module
+	quit   chan struct{}
+
+	packetsHandler func(uint64) error
+	bytesHandler   func(uint64) error
+}
+
+func (c *Controller) Stop() {
+	c.quit <- struct{}{}
+}
+
+func (c *Controller) SetPacketsHandler(h func(uint64) error) {
+	c.packetsHandler = h
+}
+func (c *Controller) SetBytesHandler(h func(uint64) error) {
+	c.bytesHandler = h
+}
+
+func (c *Controller) Run(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.quit:
+			return
+		case <-time.After(1000 * time.Millisecond):
+			lookup(c.module, packetsKey, c.packetsHandler)
+			lookup(c.module, bytesKey, c.bytesHandler)
```

Comment:
Why using a callback in parameter 3 instead of just having the `lookup` function returning the value?
```
c.packetsHandle(lookup(c.module, packetsKey))
```